File: frontend/src/core/project-workflows/canonical.ts
import type { JsonSchema, JsonValue } from "./types";
import { workflowSpecV1Schema } from "./types";

// Changing this identifier or its spelling rules is a checksum migration event.
export const CANONICAL_BINARY64_ALGORITHM =
  "ieee754-binary64-exact-decimal-v1" as const;

const binary64Buffer = new ArrayBuffer(8);
const binary64View = new DataView(binary64Buffer);
const BINARY64_FRACTION_MASK = (1n << 52n) - 1n;
const BINARY64_IMPLICIT_BIT = 1n << 52n;
const MAX_SAFE_INTEGER = (1n << 53n) - 1n;
const CANONICAL_TEXT_CHUNK_SIZE = 1_024;
const utf8Encoder = new TextEncoder();

const normalizeText = (value: string): string => {
  for (let index = 0; index < value.length; index += 1) {
    const codeUnit = value.charCodeAt(index);
    if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
      const nextCodeUnit = value.charCodeAt(index + 1);
      if (
        index + 1 >= value.length ||
        nextCodeUnit < 0xdc00 ||
        nextCodeUnit > 0xdfff
      ) {
        throw new Error(
          "canonical JSON text contains an unpaired UTF-16 surrogate",
        );
      }
      index += 1;
    } else if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
      throw new Error(
        "canonical JSON text contains an unpaired UTF-16 surrogate",
      );
    }
  }

  return value.normalize("NFC");
};

const compareText = (left: string, right: string): number => {
  const normalizedLeft = normalizeText(left);
  const normalizedRight = normalizeText(right);
  const leftCodePoints = Array.from(normalizedLeft, (character) =>
    character.codePointAt(0),
  );
  const rightCodePoints = Array.from(normalizedRight, (character) =>
    character.codePointAt(0),
  );
  const sharedLength = Math.min(leftCodePoints.length, rightCodePoints.length);
  for (let index = 0; index < sharedLength; index += 1) {
    const leftCodePoint = leftCodePoints[index]!;
    const rightCodePoint = rightCodePoints[index]!;
    if (leftCodePoint < rightCodePoint) return -1;
    if (leftCodePoint > rightCodePoint) return 1;
  }
  if (leftCodePoints.length < rightCodePoints.length) return -1;
  if (leftCodePoints.length > rightCodePoints.length) return 1;
  return 0;
};

const encodeCanonicalNumber = (value: number): string => {
  binary64View.setFloat64(0, value, false);
  const bits = binary64View.getBigUint64(0, false);
  const negative = bits >> 63n === 1n;
  const exponentBits = Number((bits >> 52n) & 0x7ffn);
  const fraction = bits & BINARY64_FRACTION_MASK;
  if (exponentBits === 0x7ff) {
    throw new Error("canonical JSON numbers must be finite");
  }
  if (exponentBits === 0 && fraction === 0n) return "0";

  let significand: bigint;
  let binaryExponent: number;
  if (exponentBits === 0) {
    significand = fraction;
    binaryExponent = -1074;
  } else {
    significand = BINARY64_IMPLICIT_BIT | fraction;
    binaryExponent = exponentBits - 1023 - 52;
  }

  if (binaryExponent >= 0) {
    const integer = significand << BigInt(binaryExponent);
    if (integer > MAX_SAFE_INTEGER) {
      throw new Error("canonical JSON integer exceeds safe-integer range");
    }
    return `${negative ? "-" : ""}${integer.toString(10)}`;
  }

  let denominatorPower = -binaryExponent;
  while (denominatorPower > 0 && (significand & 1n) === 0n) {
    significand >>= 1n;
    denominatorPower -= 1;
  }

  if (denominatorPower === 0) {
    if (significand > MAX_SAFE_INTEGER) {
      throw new Error("canonical JSON integer exceeds safe-integer range");
    }
    return `${negative ? "-" : ""}${significand.toString(10)}`;
  }

  const digits = (significand * 5n ** BigInt(denominatorPower)).toString(10);
  const scientificExponent = digits.length - 1 - denominatorPower;
  const coefficient =
    digits.length === 1 ? digits : `${digits[0]}.${digits.slice(1)}`;
  return `${negative ? "-" : ""}${coefficient}e${scientificExponent}`;
};

const canonicalizeJsonValue = (value: JsonValue): JsonValue => {
  if (typeof value === "string") return normalizeText(value);
  if (typeof value === "number") {
    encodeCanonicalNumber(value);
    return Object.is(value, -0) ? 0 : value;
  }
  if (value === null || typeof value === "boolean") return value;
  if (Array.isArray(value)) return value.map(canonicalizeJsonValue);

  const normalizedEntries = Object.entries(value)
    .filter(([, nestedValue]) => nestedValue !== undefined)
    .map(
      ([key, nestedValue]) =>
        [normalizeText(key), canonicalizeJsonValue(nestedValue)] as const,
    )
    .sort(([left], [right]) => compareText(left, right));

  for (let index = 1; index < normalizedEntries.length; index += 1) {
    if (normalizedEntries[index - 1]?.[0] === normalizedEntries[index]?.[0]) {
      throw new Error("Unicode normalization produced duplicate JSON keys");
    }
  }

  return Object.fromEntries(normalizedEntries);
};

const omitFields = (
  value: Record<string, unknown>,
  fields: ReadonlySet<string>,
): Record<string, unknown> =>
  Object.fromEntries(
    Object.entries(value).filter(([field]) => !fields.has(field)),
  );

const nodePresentationFields = new Set(["custom_label", "description"]);
const declarationPresentationFields = new Set(["description"]);

export const canonicalizeWorkflowSemanticValue = (
  input: unknown,
): JsonSchema => {
  const spec = workflowSpecV1Schema.parse(input);
  const semanticProjection = {
    schema_version: spec.schema_version,
    entry_node_id: spec.entry_node_id,
    nodes: [...spec.nodes]
      .sort((left, right) => compareText(left.id, right.id))
      .map((node) =>
        omitFields(
          node as unknown as Record<string, unknown>,
          nodePresentationFields,
        ),
      ),
    transitions: [...spec.transitions].sort((left, right) =>
      compareText(left.id, right.id),
    ),
    workflow_inputs: [...spec.workflow_inputs]
      .sort((left, right) => compareText(left.id, right.id))
      .map((declaration) =>
        omitFields(declaration, declarationPresentationFields),
      ),
    workflow_outputs: [...spec.workflow_outputs]
      .sort((left, right) => compareText(left.id, right.id))
      .map((declaration) =>
        omitFields(declaration, declarationPresentationFields),
      ),
    credential_slots: [...spec.credential_slots].sort((left, right) =>
      compareText(left.id, right.id),
    ),
  } as unknown as JsonValue;

  return canonicalizeJsonValue(semanticProjection) as JsonSchema;
};

export const serializeWorkflowSemanticChecksumInput = (
  input: unknown,
): string =>
  serializeCanonicalJsonValue(canonicalizeWorkflowSemanticValue(input));

const serializeCanonicalValue = (value: JsonValue): string => {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") return encodeCanonicalNumber(value);
  if (typeof value === "string") return JSON.stringify(normalizeText(value));
  if (Array.isArray(value)) {
    return `[${value.map(serializeCanonicalValue).join(",")}]`;
  }

  const normalizedEntries = Object.entries(value)
    .filter(([, nestedValue]) => nestedValue !== undefined)
    .map(([key, nestedValue]) => [normalizeText(key), nestedValue] as const)
    .sort(([left], [right]) => compareText(left, right));
  for (let index = 1; index < normalizedEntries.length; index += 1) {
    if (normalizedEntries[index - 1]?.[0] === normalizedEntries[index]?.[0]) {
      throw new Error("Unicode normalization produced duplicate JSON keys");
    }
  }

  return `{${normalizedEntries
    .map(
      ([key, nestedValue]) =>
        `${JSON.stringify(key)}:${serializeCanonicalValue(nestedValue)}`,
    )
    .join(",")}}`;
};

export const serializeCanonicalJsonValue = (value: JsonValue): string =>
  serializeCanonicalValue(value);

export class CanonicalJsonUtf8BudgetExceededError extends Error {
  constructor() {
    super("canonical JSON exceeds the UTF-8 byte budget");
    this.name = "CanonicalJsonUtf8BudgetExceededError";
  }
}

class CanonicalUtf8Writer {
  readonly chunks: string[] = [];
  utf8Bytes = 0;

  constructor(readonly maxUtf8Bytes: number) {
    if (!Number.isSafeInteger(maxUtf8Bytes) || maxUtf8Bytes < 0) {
      throw new Error(
        "canonical JSON UTF-8 byte budget must be a non-negative safe integer",
      );
    }
  }

  append(value: string): void {
    const byteCount = utf8Encoder.encode(value).byteLength;
    if (this.utf8Bytes + byteCount > this.maxUtf8Bytes) {
      throw new CanonicalJsonUtf8BudgetExceededError();
    }
    this.chunks.push(value);
    this.utf8Bytes += byteCount;
  }

  finish(): string {
    return this.chunks.join("");
  }
}

const jsonCharacterEscape = (character: string): string => {
  switch (character) {
    case '"':
      return '\\"';
    case "\\":
      return "\\\\";
    case "\b":
      return "\\b";
    case "\f":
      return "\\f";
    case "\n":
      return "\\n";
    case "\r":
      return "\\r";
    case "\t":
      return "\\t";
    default: {
      const codePoint = character.codePointAt(0)!;
      return codePoint < 0x20
        ? `\\u${codePoint.toString(16).padStart(4, "0")}`
        : character;
    }
  }
};

const serializeCanonicalTextWithinBudget = (
  value: string,
  writer: CanonicalUtf8Writer,
): void => {
  const normalized = normalizeText(value);
  writer.append('"');
  let buffered = "";
  for (const character of normalized) {
    buffered += jsonCharacterEscape(character);
    if (buffered.length >= CANONICAL_TEXT_CHUNK_SIZE) {
      writer.append(buffered);
      buffered = "";
    }
  }
  if (buffered.length > 0) writer.append(buffered);
  writer.append('"');
};

const serializeCanonicalValueWithinBudget = (
  value: JsonValue,
  writer: CanonicalUtf8Writer,
): void => {
  if (value === null) {
    writer.append("null");
    return;
  }
  if (typeof value === "boolean") {
    writer.append(value ? "true" : "false");
    return;
  }
  if (typeof value === "number") {
    writer.append(encodeCanonicalNumber(value));
    return;
  }
  if (typeof value === "string") {
    serializeCanonicalTextWithinBudget(value, writer);
    return;
  }
  if (Array.isArray(value)) {
    writer.append("[");
    let index = 0;
    for (const nestedValue of value) {
      if (index > 0) writer.append(",");
      serializeCanonicalValueWithinBudget(nestedValue, writer);
      index += 1;
    }
    writer.append("]");
    return;
  }

  const normalizedEntries = Object.entries(value)
    .filter(([, nestedValue]) => nestedValue !== undefined)
    .map(([key, nestedValue]) => [normalizeText(key), nestedValue] as const)
    .sort(([left], [right]) => compareText(left, right));
  for (let index = 1; index < normalizedEntries.length; index += 1) {
    if (normalizedEntries[index - 1]?.[0] === normalizedEntries[index]?.[0]) {
      throw new Error("Unicode normalization produced duplicate JSON keys");
    }
  }

  writer.append("{");
  normalizedEntries.forEach(([key, nestedValue], index) => {
    if (index > 0) writer.append(",");
    serializeCanonicalTextWithinBudget(key, writer);
    writer.append(":");
    serializeCanonicalValueWithinBudget(nestedValue, writer);
  });
  writer.append("}");
};

export const serializeCanonicalJsonValueWithinUtf8Budget = (
  value: JsonValue,
  maxUtf8Bytes: number,
): { canonical: string; utf8Bytes: number } => {
  const writer = new CanonicalUtf8Writer(maxUtf8Bytes);
  serializeCanonicalValueWithinBudget(value, writer);
  return { canonical: writer.finish(), utf8Bytes: writer.utf8Bytes };
};

File: backend/packages/harness/deerflow/workflows/canonical.py
"""Deterministic semantic serialization for WorkflowSpec v1."""

from __future__ import annotations

import hashlib
import json
import struct
import unicodedata
from typing import Any

from deerflow.workflows.contracts import (
    MAX_SAFE_JSON_INTEGER,
    JsonValue,
    WorkflowSpecV1,
    workflow_spec_public_projection_v1,
)

# Changing this identifier or its spelling rules is a checksum migration event.
CANONICAL_BINARY64_ALGORITHM = "ieee754-binary64-exact-decimal-v1"

_BINARY64_EXPONENT_MASK = 0x7FF
_BINARY64_FRACTION_MASK = (1 << 52) - 1
_BINARY64_IMPLICIT_BIT = 1 << 52
_CANONICAL_TEXT_CHUNK_SIZE = 1_024
_JSON_CHARACTER_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\f": "\\f",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
}


class CanonicalJsonUtf8BudgetExceeded(ValueError):
    """Raised before a canonical JSON value can exceed its UTF-8 budget."""


class _CanonicalUtf8Writer:
    def __init__(self, max_utf8_bytes: int) -> None:
        if type(max_utf8_bytes) is not int or max_utf8_bytes < 0:
            raise ValueError("canonical JSON UTF-8 byte budget must be a non-negative integer")
        self._max_utf8_bytes = max_utf8_bytes
        self._pieces: list[str] = []
        self.utf8_bytes = 0

    def append(self, value: str) -> None:
        byte_count = len(value.encode("utf-8"))
        if self.utf8_bytes + byte_count > self._max_utf8_bytes:
            raise CanonicalJsonUtf8BudgetExceeded("canonical JSON exceeds the UTF-8 byte budget")
        self._pieces.append(value)
        self.utf8_bytes += byte_count

    def finish(self) -> str:
        return "".join(self._pieces)


def _normalize_text(value: str) -> str:
    normalized = unicodedata.normalize("NFC", value)
    if any(0xD800 <= ord(character) <= 0xDFFF for character in normalized):
        raise ValueError("canonical Workflow JSON supports only Unicode scalar values")
    return normalized


def _canonical_number(value: int | float) -> str:
    if isinstance(value, int):
        if abs(value) > MAX_SAFE_JSON_INTEGER:
            raise ValueError("canonical Workflow JSON integer exceeds the cross-runtime safe range")
        return str(value)

    bits = int.from_bytes(struct.pack(">d", value), byteorder="big")
    negative = bool(bits >> 63)
    exponent_bits = (bits >> 52) & _BINARY64_EXPONENT_MASK
    fraction = bits & _BINARY64_FRACTION_MASK
    if exponent_bits == _BINARY64_EXPONENT_MASK:
        raise ValueError("canonical Workflow JSON does not support non-finite numbers")
    if exponent_bits == 0 and fraction == 0:
        return "0"

    if exponent_bits == 0:
        significand = fraction
        binary_exponent = -1074
    else:
        significand = _BINARY64_IMPLICIT_BIT | fraction
        binary_exponent = exponent_bits - 1023 - 52

    if binary_exponent >= 0:
        integer = significand << binary_exponent
        if integer > MAX_SAFE_JSON_INTEGER:
            raise ValueError("canonical Workflow JSON integer exceeds the cross-runtime safe range")
        return f"{'-' if negative else ''}{integer}"

    denominator_power = -binary_exponent
    trailing_zero_bits = (significand & -significand).bit_length() - 1
    common_power = min(denominator_power, trailing_zero_bits)
    significand >>= common_power
    denominator_power -= common_power

    if denominator_power == 0:
        if significand > MAX_SAFE_JSON_INTEGER:
            raise ValueError("canonical Workflow JSON integer exceeds the cross-runtime safe range")
        return f"{'-' if negative else ''}{significand}"

    digits = str(significand * 5**denominator_power)
    scientific_exponent = len(digits) - 1 - denominator_power
    coefficient = digits[0] if len(digits) == 1 else f"{digits[0]}.{digits[1:]}"
    return f"{'-' if negative else ''}{coefficient}e{scientific_exponent}"


def _canonical_encode(value: Any) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, int | float):
        return _canonical_number(value)
    if isinstance(value, str):
        return json.dumps(_normalize_text(value), ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, list):
        return "[" + ",".join(_canonical_encode(item) for item in value) + "]"
    if isinstance(value, dict):
        if not all(isinstance(key, str) for key in value):
            raise TypeError("canonical Workflow JSON object keys must be strings")
        normalized_items: list[tuple[str, Any]] = []
        normalized_keys: set[str] = set()
        for key, nested_value in value.items():
            normalized_key = _normalize_text(key)
            if normalized_key in normalized_keys:
                raise ValueError("Unicode normalization produced duplicate JSON keys")
            normalized_keys.add(normalized_key)
            normalized_items.append((normalized_key, nested_value))
        normalized_items.sort(key=lambda item: item[0])
        return "{" + ",".join(f"{_canonical_encode(key)}:{_canonical_encode(nested_value)}" for key, nested_value in normalized_items) + "}"
    raise TypeError(f"unsupported canonical Workflow JSON value: {type(value).__name__}")


def _canonical_encode_text_with_budget(value: str, writer: _CanonicalUtf8Writer) -> None:
    normalized = _normalize_text(value)
    writer.append('"')
    buffered: list[str] = []
    buffered_characters = 0
    for character in normalized:
        escaped = _JSON_CHARACTER_ESCAPES.get(character)
        if escaped is None:
            code_point = ord(character)
            escaped = f"\\u{code_point:04x}" if code_point < 0x20 else character
        buffered.append(escaped)
        buffered_characters += len(escaped)
        if buffered_characters >= _CANONICAL_TEXT_CHUNK_SIZE:
            writer.append("".join(buffered))
            buffered.clear()
            buffered_characters = 0
    if buffered:
        writer.append("".join(buffered))
    writer.append('"')


def _canonical_encode_with_budget(value: Any, writer: _CanonicalUtf8Writer) -> None:
    if value is None:
        writer.append("null")
        return
    if value is True:
        writer.append("true")
        return
    if value is False:
        writer.append("false")
        return
    if isinstance(value, int | float):
        writer.append(_canonical_number(value))
        return
    if isinstance(value, str):
        _canonical_encode_text_with_budget(value, writer)
        return
    if isinstance(value, list):
        writer.append("[")
        for index, item in enumerate(value):
            if index:
                writer.append(",")
            _canonical_encode_with_budget(item, writer)
        writer.append("]")
        return
    if isinstance(value, dict):
        if not all(isinstance(key, str) for key in value):
            raise TypeError("canonical Workflow JSON object keys must be strings")
        normalized_items: list[tuple[str, Any]] = []
        normalized_keys: set[str] = set()
        for key, nested_value in value.items():
            normalized_key = _normalize_text(key)
            if normalized_key in normalized_keys:
                raise ValueError("Unicode normalization produced duplicate JSON keys")
            normalized_keys.add(normalized_key)
            normalized_items.append((normalized_key, nested_value))
        normalized_items.sort(key=lambda item: item[0])
        writer.append("{")
        for index, (key, nested_value) in enumerate(normalized_items):
            if index:
                writer.append(",")
            _canonical_encode_text_with_budget(key, writer)
            writer.append(":")
            _canonical_encode_with_budget(nested_value, writer)
        writer.append("}")
        return
    raise TypeError(f"unsupported canonical Workflow JSON value: {type(value).__name__}")


def _sort_declarations(values: list[dict[str, Any]]) -> list[dict[str, Any]]:
    normalized_ids: set[str] = set()
    for value in values:
        normalized_id = _normalize_text(str(value.get("id", "")))
        if normalized_id in normalized_ids:
            raise ValueError("Unicode normalization produced duplicate declaration IDs")
        normalized_ids.add(normalized_id)
    return sorted(values, key=lambda value: _normalize_text(str(value.get("id", ""))))


def semantic_payload(spec: WorkflowSpecV1) -> dict[str, Any]:
    """Return the execution-semantic v1 projection used for checksums.

    Only top-level declaration arrays are order-normalized.  Nested arrays are
    kept byte-for-byte in authored order because branch order, aggregate
    candidates, loop variables, templates, and similar arrays may be semantic.
    """

    payload = workflow_spec_public_projection_v1(spec)

    nodes = payload["nodes"]
    for node in nodes:
        node.pop("custom_label", None)
        node.pop("description", None)
    payload["nodes"] = _sort_declarations(nodes)

    transitions = payload["transitions"]
    payload["transitions"] = _sort_declarations(transitions)

    workflow_inputs = payload["workflow_inputs"]
    for declaration in workflow_inputs:
        declaration.pop("description", None)
    payload["workflow_inputs"] = _sort_declarations(workflow_inputs)

    workflow_outputs = payload["workflow_outputs"]
    for declaration in workflow_outputs:
        declaration.pop("description", None)
    payload["workflow_outputs"] = _sort_declarations(workflow_outputs)

    payload["credential_slots"] = _sort_declarations(payload["credential_slots"])
    return payload


def semantic_canonical_json(spec: WorkflowSpecV1) -> str:
    """Serialize a WorkflowSpec to its fixed, Unicode-preserving JSON form."""

    return _canonical_encode(semantic_payload(spec))


def canonical_json_value(value: JsonValue) -> str:
    """Serialize one portable JSON value using the Workflow canonical form."""

    return _canonical_encode(value)


def canonical_json_value_with_utf8_budget(value: JsonValue, *, max_utf8_bytes: int) -> tuple[str, int]:
    """Serialize canonical JSON while stopping as soon as its UTF-8 budget is crossed."""

    writer = _CanonicalUtf8Writer(max_utf8_bytes)
    _canonical_encode_with_budget(value, writer)
    return writer.finish(), writer.utf8_bytes


def semantic_checksum(spec: WorkflowSpecV1) -> str:
    """Return the lowercase SHA-256 digest of semantic canonical UTF-8 JSON."""

    canonical = semantic_canonical_json(spec)
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()
